File: estructuras/ListaSecuencial.py
import numpy as np


class ListaSecuencial:
    __dimension: int
    __ultimo: int
    __cantidad: int
    __items: np.ndarray

    def __init__(self, dimension):
        self.__dimension = dimension
        self.__cantidad = 0
        self.__ultimo = -1
        self.__items = np.empty(dimension, dtype=int)

    def insertar_posicion(self, item, posicion):
        if self.__cantidad == self.__dimension:
            print("Lista llena")
            return None
        if 0 <= posicion < self.__dimension and self.hay_posicion_disponible():
            i = self.__ultimo + 1
            while i > posicion:
                self.__items[i] = self.__items[i - 1]
                i -= 1
            self.__items[i] = item
            self.__cantidad += 1
            self.__ultimo += 1
            return item
        print("Posicion invalida")
        return None

    def insertar_contenido(self, item):
        if self.__cantidad == self.__dimension:
            print("Lista llena")
            return None
        if self.vacia():
            self.__ultimo += 1
            self.__items[self.__ultimo] = item
            self.__cantidad += 1
            return item
        i = 0
        while i <= self.__ultimo and self.__items[i] < item:
            i += 1
        for j in range(self.__ultimo + 1, i, -1):
            self.__items[j] = self.__items[j - 1]
        self.__items[i] = item
        self.__ultimo += 1
        self.__cantidad += 1
        return item

    def suprimir_posicion(self, posicion):
        if self.vacia():
            print("Lista vacia")
            return None
        if 0 <= posicion <= self.__ultimo:
            i = self.__ultimo
            eliminado = self.__items[posicion]
            while i < posicion:
                self.__items[i] = self.__items[i + 1]
                i += 1
            self.__ultimo -= 1
            self.__cantidad -= 1
            return eliminado
        print("Posicion invalida")
        return None

    def suprimir_contenido(self, item):
        if self.vacia():
            print("Lista vacia")
            return None
        i = 0
        while i <= self.__ultimo and self.__items[i] < item:
            i += 1
        if self.__items[i] != item:
            print("Elemento no encontrado")
            return None
        eliminado = self.__items[i]
        for j in range(i, self.__ultimo):
            self.__items[j] = self.__items[j + 1]
        self.__ultimo -= 1
        self.__cantidad -= 1
        return eliminado

    def buscar(self, item):
        if self.vacia():
            print("Lista vacia")
            return None

        i = 0
        while i <= self.__ultimo and self.__Items[i] < item:
            i += 1
        if self.__items[i] == item:
            return i
        print("Elemento no encontrado")
        return None

    def primer_elemento(self):
        if self.vacia():
            print("Lista vacia")
            return None
        return self.__items[0]

    def ultimo_elemento(self):
        if self.vacia():
            print("Lista vacia")
            return None
        return self.__items[self.__ultimo]

    def siguiente(self, posicion):
        if 0 <= posicion <= self.__ultimo:
            if posicion == self.__ultimo:
                print("Ingreso la ultima posicion")
                return None
            return self.__items[posicion + 1]
        print("Posicion invalida")
        return None

    def anterior(self, posicion):
        if 0 <= posicion <= self.__ultimo:
            if posicion == 0:
                print("Ingreso la primera posicion")
                return None
            return self.__items[posicion - 1]
        print("Posicion invalida")
        return None

    def recorrer(self):
        if self.vacia():
            print("Lista vacia")
            return None
        for i in range(self.__ultimo + 1):
            print(self.__items[i], end=" ")
        print()

    def vacia(self):
        return self.__cantidad == 0

    def llena(self):
        return self.__cantidad == self.__dimension

File: csharp/Lista/Lista/ListaSecuencial.cs
using System.Globalization;

namespace Lista
{
  class Secuencial(int dim)
  {
    private readonly int Dimension = dim;
    private int Cantidad = 0;
    private int[] Items = new int[dim];
    private int Ultimo = -1;

    public int InsertarContenido(int item)
    {
      if (Cantidad == Dimension)
      {
        Console.WriteLine("Lista llena");
        return -1;
      }
      int i = 0;
      while (i <= Ultimo && Items[i] < item)
      {
        i++;
      }
      for (int j = Ultimo; j > i; j--)
      {
        Items[j] = Items[j - 1];
      }
      Items[i] = item;
      Cantidad++;
      Ultimo++;
      return Items[i];
    }

    public int InsertarPosicion(int item, int posicion)
    {
      if (Cantidad == Dimension)
      {
        Console.WriteLine("Lista llena");
        return -1;
      }
      if (0 <= posicion && posicion <= Ultimo)
      {
        int i = Ultimo + 1;
        while (i > posicion)
        {
          Items[i] = Items[i - 1];
          i--;
        }
        Items[posicion] = item;
        Cantidad++;
        Ultimo++;
        return Items[posicion];
      }
      Console.WriteLine("Posicion no valida");
      return -1;
    }

    public int SuprimirContenido(int item)
    {
      if (Vacia())
      {
        Console.WriteLine("Lista vacia.");
        return -1;
      }
      int i = 0;
      while (i <= Ultimo && Items[i] == item)
      {
        i++;
      }
      if (Items[i] != item)
      {
        Console.WriteLine("Item no valido");
        return -1;
      }
      int eliminado = Items[i];
      for (int j = i; j <= Ultimo; j++)
      {
        Items[j] = Items[j + 1];
      }
      Cantidad--;
      Ultimo--;
      return eliminado;
    }

    public int SuprimirPosicion(int posicion)
    {
      if (Vacia())
      {
        Console.WriteLine("Lista vacia.");
        return -1;
      }
      if (0 <= posicion && posicion <= Ultimo)
      {
        int eliminado = Items[posicion];
        while (posicion <= Ultimo)
        {
          Items[posicion] = Items[posicion++];
        }
        Cantidad--;
        Ultimo--;
        return eliminado;
      }
      Console.WriteLine("Posicion no valida");
      return -1;
    }

    public int Buscar(int item)
    {
      if (Vacia())
      {
        Console.WriteLine("Lista vacia.");
        return -1;
      }
      int i = 0;
      while (i <= Ultimo && Items[i] != item)
      {
        i++;
      }
      if (Items[i] == item)
      {
        return item;
      }
      Console.WriteLine("Elemento no valido");
      return -1;
    }

    public void Recorrer()
    {
      for (int i = 0; i < Cantidad; i++)
      {
        Console.Write(Items[i] + " ");
      }
      Console.WriteLine();
    }

    public int PrimerElemento()
    {
      if (Vacia())
      {
        Console.WriteLine("Lista vacia.");
        return -1;
      }
      return Items[0];
    }

    public int UltimoElemento()
    {
      if (Vacia())
      {
        Console.WriteLine("Lista vacia.");
        return -1;
      }
      return Items[Ultimo];
    }

    public int Siguiente(int posicion)
    {

      if (Vacia())
      {
        Console.WriteLine("Lista vacia.");
        return -1;
      }
      if (posicion == Ultimo)
      {
        Console.WriteLine("Ingreso la ultima posicion");
        return -1;
      }
      return Items[posicion + 1];
    }

    public int Anterior(int posicion)
    {

      if (Vacia())
      {
        Console.WriteLine("Lista vacia.");
        return -1;
      }
      if (posicion == 0)
      {
        Console.WriteLine("Ingreso la primera posicion");
        return -1;
      }
      return Items[posicion - 1];
    }

    public bool Vacia()
    {
      return Cantidad == 0;
    }

    public bool Llena()
    {
      return Cantidad == Dimension;
    }
  }

}